libs/content/workspace/src/task_manager.rs

```diff
@@ -0,0 +1,462 @@
+
+        // Syncs don't need to be prioritized because they don't conflict with each other or with loads or saves. For
+        // efficiency, we wait for all saves to complete before we launch a sync. A save always queues a sync upon
+        // completion.
+        let should_sync = !self.queued_syncs.is_empty()
+            && self.in_progress_sync.is_none()
+            && !self.any_load_or_save_queued_or_in_progress();
+
+        // Get launched things from the queue and remove duplicates (avoiding clones)
+        let mut loads_to_launch = HashMap::new();
+        let mut saves_to_launch = HashMap::new();
+
+        for load in mem::take(&mut self.queued_loads) {
+            if ids_to_load.contains(&load.request.id) {
+                loads_to_launch.insert(load.request.id, load); // use latest of duplicate ids
+            } else {
+                self.queued_loads.push(load); // put back the ones we're not launching
+            }
+        }
+        for save in mem::take(&mut self.queued_saves) {
+            if ids_to_save.contains(&save.request.id) {
+                saves_to_launch.insert(save.request.id, save); // use latest of duplicate ids
+            } else {
+                self.queued_saves.push(save); // put back the ones we're not launching
+            }
+        }
+
+        let sync_to_launch =
+            if should_sync { mem::take(&mut self.queued_syncs).into_iter().next() } else { None };
+
+        // Actual launching happens in `impl FileCacheExt for Arc<Mutex<FileCache>>` block because it needs cloned Arc
+        InnerResponse {
+            loads_to_launch: loads_to_launch.into_values().collect(),
+            saves_to_launch: saves_to_launch.into_values().collect(),
+            sync_to_launch,
+            completed_loads: mem::take(&mut self.completed_loads),
+            completed_saves: mem::take(&mut self.completed_saves),
+            completed_sync: mem::take(&mut self.completed_sync),
+        }
+    }
+
+    fn load_or_save_in_progress(&self, id: Uuid) -> bool {
+        let save_in_progress = self
+            .in_progress_saves
+            .iter()
+            .any(|in_progress_save| in_progress_save.request.id == id);
+        let load_in_progress = self
+            .in_progress_loads
+            .iter()
+            .any(|in_progress_load| in_progress_load.request.id == id);
+        save_in_progress || load_in_progress
+    }
+
+    fn any_load_or_save_queued_or_in_progress(&self) -> bool {
+        !self.queued_loads.is_empty()
+            || !self.queued_saves.is_empty()
+            || !self.in_progress_loads.is_empty()
+            || !self.in_progress_saves.is_empty()
+    }
+}
+
+pub trait FileCacheExt {
+    // pass-throughs
+    fn new() -> Self;
+    fn queue_load(&mut self, request: LoadRequest);
+    fn queue_save(&mut self, request: SaveRequest);
+    fn queue_sync(&mut self);
+
+    // non-pass-throughs
+    fn update(&mut self, ctx: &Context, core: &Lb) -> Response;
+}
```

Comment:
This way makes it less noisy to use the TaskManager because you don't have to reference the inner field to lock the mutex, but no strong feelings, I'll just change it if you prefer it that way